nixos/modules/services/networking/adguard-dnsproxy.nix

```diff
@@ -0,0 +1,98 @@
+{ config, lib, pkgs, ... }:
+with lib;
+
+let cfg = config.services.adguard-dnsproxy;
+
+in {
+  meta.maintainers = with lib.maintainers; [ ckie ];
+  options.services.adguard-dnsproxy = {
+    enable = mkEnableOption
+      "dnsproxy, a simple DNS proxy with DoH, DoT, DoQ and DNSCrypt support";
+    extraOptions = mkOption {
+      type = types.listOf types.str;
+      default = [ ];
+      description = "Extra arguments to pass to dnsproxy";
+    };
+    package = mkOption {
+      type = types.package;
+      default = pkgs.dnsproxy;
+      defaultText = literalExpression "pkgs.dnsproxy";
+      description = "The package containing the dnsproxy binary";
+    };
+    port = mkOption {
+      type = types.nullOr types.port;
+      default = null;
+      description = "An optional port to listen on for raw DNS traffic";
+    };
+    upstreams = mkOption {
+      type = types.listOf types.str;
+      default = null;
+      description = "Upstream servers to talk with";
+    };
+    fallbacks = mkOption {
+      type = types.listOf types.str;
+      default = [ "1.1.1.1" "9.9.9.9" "8.8.8.8" ];
+      description =
+        "Fallback upstream servers to talk with, used for bootstrapping the connection to the primary upstreams";
+    };
+    verbose = mkOption {
+      type = types.bool;
+      default = false;
+      description = "Enable verbose debugging output";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    services.adguard-dnsproxy.extraOptions = [ ]
+      ++ optional (cfg.port != null) "--port ${toString cfg.port}"
+      ++ map (u: "--upstream ${escapeShellArg u}") cfg.upstreams
+      ++ map (u: "--fallback ${escapeShellArg u}") cfg.fallbacks
+      ++ optional cfg.verbose "--verbose";
+
+    systemd.services.adguard-dnsproxy = {
+      description = "adguard-dnsproxy";
+      wants = [ "network-online.target" "nss-lookup.target" ];
+      before = [ "nss-lookup.target" ];
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
```

Comment:
Where do these hardening options come from? Upstream documentation? Experimentation?

Reply:
Oh, I just stole them from the `dnscrypt-proxy2` module: systemd seems to be happy and it still runs.

Reply:
Sounds good. Thanks.